File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

long long solve(long long level, int diff, int prev, int cur) {
    if (level >= diff) return cur;
    
    return (diff - level) * (prev + cur) + cur;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    int len = times.size();
    
    long long maxValue = -1;
    long long left = 1;
    long long right = limit;
    

    
    while (left <= right) {
        long long mid = left + (right - left) / 2;
        
        long long result = times[0];
        for (int i = 1; i < len; i++) {
            result += solve(mid, diffs[i], times[i - 1], times[i]);
        }

        
        if (result <= limit) {
            maxValue = mid;
            right = mid -1;
        } else {
            left = mid + 1;
        }
    }
    answer = maxValue;
    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 0
    left = 1
    right = 100000
    length = len(diffs)
    
    while (left <= right):
        level = (left + right) // 2
        
        
        # diff <= level -> time_cur
        # diff > level -> (diff - level) * (time_cur + time_prev) + time_cur
        
        solved = 0
        for i in range(length):
            diff = diffs[i]
            time = times[i]
            time_prev = times[i - 1] if i > 0 else 0
            
            if (diff <= level):
                solved += time
            elif (diff > level):
                solved += (diff - level) * (time + time_prev) + time
                
            if (solved > limit):
                break
        if (solved > limit):
            left = level + 1
        elif (solved <= limit):
            answer = level
            right = level - 1
            
    
    return answer